Enable value types escape analysis

hzongaro opened the pull request:
These changes enable support for stack allocation of instances of value type classes in escape analysis.  The changes are staged through a sequence of commits to aid in reviewing:

1. Commit 16d4e5ec4ac4d9bc390ebb2c8fb2229de283abb2 adds support for contiguous stack allocation of instances of value type classes
2. Commit 075335a1a65d53146590de54071cd3a6c6c91e14 adds support for non-contiguous stack allocation of instances of value type classes
3. Commit c2951330f46ea107e6d401e1154d51d3bedfeeef adds support for cold block heapification of instances of value type classes
4. Commit 404cd86f5cd9c6c651c3eee86b608cf104c7df86 adds special "harmless" recognition of array element load operations through calls to helper or non-helper methods
5. Commit 453a61a077717d12b4463a984cb1e872b4db8c66 adds some unit tests of escape analysis of value type instances
6. Commit 9e727ec12d64b65b38afe223fec21ab4615810dc is a minor tracing improvement

This pull request is currently marked as a draft pull request as it depends on a downstream OMR pull request.  However, I am not expecting to make any further changes to this pull request, except as might be required to address review comments, of course.

Depends:  eclipse/omr#7031

a7ehuo commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6674,13 +6722,86 @@ void TR_EscapeAnalysis::makeContiguousLocalAllocation(Candidate *candidate)
+      const TR::TypeLayout* typeLayout = comp()->typeLayout(valueClass);
+
+      TR::TreeTop *fieldValueTreeTopCursor = candidate->_treeTop->getPrevTreeTop();
+      TR::TreeTop *fieldStoreTreeTopCursor = candidate->_treeTop;
```
Could an example of IL trees be added to help others understand what stack allocated `newvalue` trees look like compared to the origin `newvalue` IL tree?

hzongaro replied:
Fixed in commit 7ebca667516f0962fcb3fd7ab40e5501c1885389

a7ehuo reviewed (commented):
Here are the second half of the review comments. Just some very minor comments

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -1567,6 +1593,9 @@ void TR_EscapeAnalysis::findCandidates()
          continue;
          }
 
+      // TODO:  If java.lang.Integer is ever made into a primitive value type class, will
```
How about "TODO-VALH"?

a7ehuo replied:
That sounds good. Other alternatives could be `TODO-VT`, `TODO-VALUETYPE`, or `TODO-VALUE`. We haven't referred to value types in the JIT code as Valhalla much. I wonder keeping it as value types ish would make it easier to search code related to value types overall.

hzongaro replied:
Changed in commit 7eb3eec9fceb42599c57a802270dc60b2aaf15da

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -219,6 +229,11 @@ int32_t TR_EscapeAnalysis::perform()
 
+   if (trace())
+      {
+      traceMsg(comp(), "In Escape Analysis for %s\n", comp()->signature());
```
I can't recall now how I came to add it.  There must have been a situation where other debug output came before the point where trees were being dumped in [`EscapeAnalysis::performAnalysisOnce`](https://github.com/eclipse-openj9/openj9/blob/16d4e5ec4ac4d9bc390ebb2c8fb2229de283abb2/runtime/compiler/optimizer/EscapeAnalysis.cpp#L647), and I was trying to tell where debug output from some earlier optimization ended and output related to escape analysis began.

I'll remove this trace message for now.  If I come across such a situation again, I will add it back in a future pull request.

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6692,63 +6832,66 @@ void TR_EscapeAnalysis::makeContiguousLocalAllocation(Candidate *candidate)
 
    if (candidate->isExplicitlyInitialized())
       {
-      // Find all the explicit zero-initializations and see if any of the
-      // generic int shadows can be replaced by real field references.
+      if (originalAllocationOpCode != TR::newvalue)
```
I can do that.  There's also an `else` that's associated with the `if (candidate->isExplicitlyInitialized())`, so I'll have to change the entire thing from this form

```
if (candidate->isExplicitlyInitialized())
   {
   if (originalAllocationOpCode != TR::newvalue)
   ...
   }
else
   ...
```

to this form

```
if (candidate->isExplicitlyInitialized() && (originalAllocationOpCode != TR::newvalue))
   ...
   }
else if (!candidate->isExplicitlyInitialized())
   ...
```

a7ehuo replied:
Now I see.  Thanks for pointing it out. On a second thought, it's probably better to keep the change as it is so that it's easier to follow the logic. When I was reviewing this part of the code, I was more thinking of how easy it is to compare the change to the previous version or existing master branch. However, now I think what's more important is how easy it is to follow the logic with this change.

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7087,7 +7234,7 @@ void TR_EscapeAnalysis::heapifyForColdBlocks(Candidate *candidate)
 
       if (trace())
-         traceMsg(comp(), "heapifying %p b1 %d b2 %d\n", candidate->_node, candidate->isContiguousAllocation(), candidate->_dememoizedMethodSymRef);
+         traceMsg(comp(), "heapifying %p:  isContiguousAllocation == %d; _dememoizedMethodSymRef %d; new heapAllocation treetop %p\n", candidate->_node, candidate->isContiguousAllocation(), candidate->_dememoizedMethodSymRef, heapAllocation);
```
I'll remove the `==`

hzongaro replied:
Changed in commit 7ebca667516f0962fcb3fd7ab40e5501c1885389

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7071,10 +7065,62 @@ void TR_EscapeAnalysis::makeNonContiguousLocalAllocation(Candidate *candidate)
+                   && (autoField._offset == fieldEntry._offset))
+                  {
+                  TR::DataType type = autoField._symRef->getSymbol()->getDataType();
+                  TR::Node *storeNode = TR::Node::createWithSymRef(comp()->il.opCodeForDirectStore(type), 1, 1, fieldValueNode, autoField._symRef);
```
No, in the non-contiguous case escape analysis stores each field's value to a local temporary, so no write barrier or compressed reference is used.  Compare this with the [similar code below](https://github.com/eclipse-openj9/openj9/blob/075335a1a65d53146590de54071cd3a6c6c91e14/runtime/compiler/optimizer/EscapeAnalysis.cpp#L7147-L7150) that handles non-contiguous `TR::new`.

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7071,10 +7065,62 @@ void TR_EscapeAnalysis::makeNonContiguousLocalAllocation(Candidate *candidate)
+                   && (autoField._offset == fieldEntry._offset))
+                  {
+                  TR::DataType type = autoField._symRef->getSymbol()->getDataType();
+                  TR::Node *storeNode = TR::Node::createWithSymRef(comp()->il.opCodeForDirectStore(type), 1, 1, fieldValueNode, autoField._symRef);
```
Just to make sure I'm understanding this comment correctly, are you suggesting that [the code in `makeNonContiguousAllocation`](https://github.com/hzongaro/openj9/blob/9e727ec12d64b65b38afe223fec21ab4615810dc/runtime/compiler/optimizer/EscapeAnalysis.cpp#L7059-L7116) that handles `TR::newvalue` should be pulled into a new method along with [the similar code from `makeContiguousAllocation`](https://github.com/hzongaro/openj9/blob/9e727ec12d64b65b38afe223fec21ab4615810dc/runtime/compiler/optimizer/EscapeAnalysis.cpp#L6740-L6802)?  Or are you suggesting that all of `makeNonContiguousLocalAllocation` and `makeContiguousLocalAllocation` could be combined?

hzongaro replied:
Fixed in commit 44e247d4add232357092ac05a2c5c9e9680ef6c9

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7361,6 +7352,9 @@ void TR_EscapeAnalysis::heapifyForColdBlocks(Candidate *candidate)
       heapStoreTree->join(heapAllocationExitTree);
 
+      TR_ByteCodeInfo bcInfo = candidate->_node->getByteCodeInfo();
+      TR::DebugCounter::prependDebugCounter(comp(), TR::DebugCounter::debugCounterName(comp(), "escapeAnalysis/heapification/%s/%s/(%s)/(%d,%d)", comp()->getHotnessName(comp()->getMethodHotness()), comp()->signature(), bcInfo.getCallerIndex() > -1 ? comp()->getInlinedResolvedMethod(bcInfo.getCallerIndex())->signature(trMemory()): comp()->signature(), bcInfo.getCallerIndex(), bcInfo.getByteCodeInd ...
```
I seem to have acquired a habit of placing all the arguments to `traceMsg` and `debugCounterName` on a single line.  I'll just make sure the format string is on the same line as the `debugCounterName` method name and all the other arguments can go on subsequent lines.

> I wonder if `bcInfo.getCallerIndex()` and `bcInfo.getByteCodeIndex()` could be moved right after `getHotnessName`.

I know the counter names can get very unwieldy.  The way debug counts are accumulated, it's useful to have less specific locations earlier in the counter name's hierarchy and more specific locations later.  I've found it useful to track how much stack allocation happens at `hot` versus `scorching`, how much happens in a particular method in its entirety at that optimization level, how much happens in code that was inlined from another method (possibly several times) into that method, and finally at specific locations in the bytecode in that method.  That's why the bytecode info appears last.

a7ehuo replied:
>I know the counter names can get very unwieldy. The way debug counts are accumulated, it's useful to have less specific locations earlier in the counter name's hierarchy and more specific locations later. I've found it useful to track how much stack allocation happens at hot versus scorching, how much happens in a particular method in its entirety at that optimization level, how much happens in code that was inlined from another method (possibly several times) into that method, and finally at specific locations in the bytecode in that method. That's why the bytecode info appears last.

Thanks for the explanation! I haven't looked at debug counter for a while. It makes sense. Please keep the order as it is.

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7097,24 +7077,27 @@ void TR_EscapeAnalysis::makeNonContiguousLocalAllocation(Candidate *candidate)
             }
+
+         TR::DataType type = autoSymRef->getSymbol()->getDataType();
+         TR::Node *storeNode = TR::Node::createWithSymRef(comp()->il.opCodeForDirectStore(type), 1, 1, fieldValueNode, autoSymRef);
```
No, in the non-contiguous case escape analysis stores each field's value to a local temporary, so no write barrier or compressed reference is used. Compare this with the [similar code below](https://github.com/eclipse-openj9/openj9/blob/075335a1a65d53146590de54071cd3a6c6c91e14/runtime/compiler/optimizer/EscapeAnalysis.cpp#L7147-L7150) that handles non-contiguous `TR::new`.

hzongaro commented on test/functional/Valhalla/src/org/openj9/test/lworld/NullRestrictedTypeOptTests.java:
```diff
@@ -0,0 +1,341 @@
+
+	@Test(priority=1)
+	static public void testEAOpportunity10() throws Throwable {
+	        for (int i = 0; i < 10000; i++) {
+	                result = 0;
+	                for (int j = 0; j < 100; j++) {
+	                        evalTestEA10a(j);
+	                }
+	                assertEquals(result, 1500);
+
+	                result = 0;
+	                evalTestEA10b();
+	                assertEquals(result, 1500);
+	        }
```
Oops! Looks like I used spaces instead of tabs.

hzongaro replied:
Corrected in commit d45810c39c379279ad8d2fe6828624a3d581460b

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6788,6 +6748,18 @@ void TR_EscapeAnalysis::makeContiguousLocalAllocation(Candidate *candidate)
                                                                           fieldEntry._typeSignature
                                                                           );
 
+         int32_t fieldSize = TR::Symbol::convertTypeToSize(fieldEntry._datatype);
+
+         if (fieldEntry._datatype.getDataType() == TR::Address && comp()->useCompressedPointers())
+            {
+            fieldSize = TR::Compiler->om.sizeofReferenceField();
+            }
```
Good point.  This appears many times in existing code in Escape Analysis and in many places throughout the JIT compiler that are unrelated to Escape Analysis.  Would you mind if I opened a follow on clean-up issue to handle that, rather than dealing with it in this pull request?

a7ehuo replied:
Opening a clean-up issue sounds good. Thank you!

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6961,16 +7151,43 @@ void TR_EscapeAnalysis::makeNonContiguousLocalAllocation(Candidate *candidate)
+         candidate->_node->setAndIncChild(0, classNode);
+         TR::Node::recreate(candidate->_node, TR::New);
+         candidate->_node->setNumChildren(1);
+         candidate->_class = classObject;
```
In making changes to address review comments, I noticed that this change to `candidate->_class` for non-contiguous value type candidates causes problems later in [`heapifyForColdBlocks`](https://github.com/eclipse-openj9/openj9/blob/9e727ec12d64b65b38afe223fec21ab4615810dc/runtime/compiler/optimizer/EscapeAnalysis.cpp#L7500-L7501) when the `_class` field is used to lookup information about the fields of the class.  I will introduce an `_origClass` field to avoid that problem.

hzongaro replied:
Fixed in commit fbf9cb620407fe4b1360b4f4603209c253d61535

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7371,117 +7365,167 @@ void TR_EscapeAnalysis::heapifyForColdBlocks(Candidate *candidate)
-//             }
+            TR_ASSERT_FATAL(stackFieldLoad != NULL, "newvalue candidate %p was missing an entry in candidate->_fields for its field number %d at offset %d", candidate->_node, idx, fieldEntry._offset);
+            newValueHeapificationNode->setAndIncChild(idx+1, stackFieldLoad);
+            traceMsg(comp(), "Updating heap allocation child %d with field load node %p\n", idx+1, stackFieldLoad);
```
Fixed in commit 7ebca667516f0962fcb3fd7ab40e5501c1885389

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7371,117 +7365,167 @@ void TR_EscapeAnalysis::heapifyForColdBlocks(Candidate *candidate)
+         TR_ASSERT_FATAL(candidate->_fields != NULL || fieldCount == 0, "candidate %p of kind newvalue is expected to have non-NULL candidate->_fields if it has a non-zero fieldCount %d\n", candidate->_node, fieldCount);
+
+         if (trace())
+            traceMsg(comp(), "Updating heap allocation node with fieldCount == %d\n", fieldCount);
```
Fixed in commit 7ebca667516f0962fcb3fd7ab40e5501c1885389

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -3757,6 +3749,25 @@ void FieldInfo::rememberFieldSymRef(TR::Node *fieldNode, int32_t fieldOffset, Ca
+      {
+      // Nothing to do
+      }
+   else
```
Fixed in commit 7ebca667516f0962fcb3fd7ab40e5501c1885389

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7102,6 +7249,8 @@ void TR_EscapeAnalysis::heapifyForColdBlocks(Candidate *candidate)
                FieldInfo &field = candidate->_fields->element(j);
                fieldSize = field._size;
+
+               // TODO:  Need to handle dememoized Integer.valueOf when Integer becomes a value type
```
Fixed in commit 7eb3eec9fceb42599c57a802270dc60b2aaf15da

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6674,13 +6753,74 @@ void TR_EscapeAnalysis::makeContiguousLocalAllocation(Candidate *candidate)
+
+         const TR::ILOpCodes storeOpCode = (fieldValueNode->getDataType() != TR::Address)
+                                              ? comp()->il.opCodeForIndirectStore(fieldValueNode->getDataType())
+                                              : comp()->il.opCodeForIndirectWriteBarrier(fieldValueNode->getDataType());
```
Fixed in commit 44e247d4add232357092ac05a2c5c9e9680ef6c9

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6674,13 +6753,74 @@ void TR_EscapeAnalysis::makeContiguousLocalAllocation(Candidate *candidate)
 
+   if (candidate->isExplicitlyInitialized() && (originalAllocationOpCode == TR::newvalue))
+      {
+      // TODO:  Can this code be commoned up with code in lowerNewValue?
```
Fixed in commit 44e247d4add232357092ac05a2c5c9e9680ef6c9

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6619,7 +6696,8 @@ void TR_EscapeAnalysis::avoidStringCopyAllocation(Candidate *candidate)
 bool TR_EscapeAnalysis::tryToZeroInitializeUsingArrayset(Candidate* candidate, TR::TreeTop* precedingTreeTop)
    {
-   if (cg()->getSupportsArraySet())
+   // TODO:  Investigate whether arrayset can be used for newvalue
```
Fixed in commit 7eb3eec9fceb42599c57a802270dc60b2aaf15da

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -4295,6 +4358,15 @@ void TR_EscapeAnalysis::checkEscapeViaNonCall(TR::Node *node, TR::NodeChecklist&
+         forceEscape(node->getChild(i), node);
+         }
+      return;
+      }
```
Fixed in commit 7eb3eec9fceb42599c57a802270dc60b2aaf15da

hzongaro commented:
Annabelle @a7ehuo, I have addressed your review comments.  I split the changes into several distinct commits to make it easier to squash each set of changes into the original set of commits that they affected.

I'll mark this pull request as ready for review now that OMR pull request eclipse/omr#7031 has been promoted into OpenJ9-OMR.

a7ehuo commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6698,6 +6698,149 @@ bool TR_EscapeAnalysis::tryToZeroInitializeUsingArrayset(Candidate* candidate, T
+            }
+
+         // if storing a ref, make sure it is compressed
+         if (comp()->useCompressedPointers() && fieldValueNode->getDataType() == TR::Address)
+            {
+            completeFieldStoreNode = TR::Node::createCompressedRefsAnchor(storeNode);
+            }
+         else
+            {
+            completeFieldStoreNode = storeNode;
+            }
```
The question here is just to help me understand the code better: With my understanding of how [non contiguous](https://github.com/eclipse-openj9/openj9/pull/17594#discussion_r1235793546) doesn't need to consider compressed ref, contiguous stack allocation also stores the field value into an offset of a local temporary (no heap access). Why would compressed ref be needed in contiguous case?

hzongaro replied:
In general a contiguously stack allocated object might be accessed at a point in the code that might also be able to access a heap allocated object.  Consider a case like the following, where class `A` has a reference field `r`.  After the `if` construct, `a.r` might be referring to a field in the object created at the line marked `Obj 1` - which must be allocated on the heap, because it escapes - or a field in the object created at the line marked `Obj 2` - which might be stack allocated, because it doesn't escape.  So the offsets and sizes of the shadow symbols used to access the fields in the contiguously stack allocated object have to be the same as those used to access a heap allocated instance of the same class.

```
static A s = new A();  // Obj 1
A a = new A();   // Obj 2
if (some-condition) {
    a = s;
}
Object o = a.r;   // Obj 1 or 2?
```

a7ehuo reviewed (commented):
LGTM. Thank you for making these updates Henry! Just two questions to help me understand the code and one very minor formatting comment in the test code

hzongaro commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -6698,6 +6698,149 @@ bool TR_EscapeAnalysis::tryToZeroInitializeUsingArrayset(Candidate* candidate, T
+   //    n119n         iconst 2 (X!=0 X>=0 )
+   //
+   // and after, for contiguous stack allocation, where #439 is aggregate storage on the stack.
+   // The field holding the vft for the object will be inserted in the method entry block.
```
My understanding is that it's because GC considers stack allocated objects in the process of garbage collection.  Stack allocated objects are considered to be live throughout the method, so the vft needs to be stored in object at method entry, and any reference fields in the object need to be zero initialized at method entry as well, so GC handles the objects (and anything they reference) correctly.

hzongaro replied:
My understanding is that it's because GC considers stack allocated objects in the process of garbage collection. Stack allocated objects are considered to be live throughout the method, so the vft needs to be stored in the object header at method entry, and any reference fields in the object need to be zero initialized at method entry as well, so GC handles the objects (and anything they reference) correctly.

a7ehuo replied:
Thank you for the clarification Henry!

hzongaro commented on test/functional/Valhalla/src/org/openj9/test/lworld/NullRestrictedTypeOptTests.java:
```diff
@@ -176,21 +181,21 @@ public NestedPrimPair(PrimPair p1, PrimPair p2) {
+
+		        result = 0;
+		        evalTestEA4b();
+		        assertEquals(result, 1500);
```
Fixed in commit a7f81f16d9d57e7f95e033edab8bf775a29eb115

a7ehuo reviewed (approved):
LGTM

a7ehuo commented on test/functional/Valhalla/src/org/openj9/test/lworld/NullRestrictedTypeOptTests.java:
```diff
@@ -0,0 +1,341 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2022
```
Sorry I was just updating other test files and noticed the copyright year. Should this be `and others 2023`?

hzongaro replied:
Good point!

a7ehuo replied:
LGTM

vijaysun-omr commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -3237,7 +3295,11 @@ void TR_EscapeAnalysis::forceEscape(TR::Node *node, TR::Node *reason, bool force
             else
+               {
+               if (trace())
+                  traceMsg(comp(), "  Marking candidate [%p] as referenced in forceEscape as it is immutable\n", candidate->_node);
```
What does this mean exactly ? i.e. what does immutability have to do with marking the candidate as referenced ?

hzongaro replied:
> what does immutability have to do with marking the candidate as referenced

I was adding more tracing to make it easier to track how a candidate comes to have some of the flags set on it.  Perhaps a better way of describing this situation would be as follows:

```
traceMsg(comp(), "  Marking immutable candidate [%p] as referenced in forceEscape to allow for non-contiguous allocation, but compensating for escape at [%p]\n", candidate->_node, reason);
```

Thoughts?

vijaysun-omr replied:
Sure that sounds better, thanks

hzongaro replied:
Updated in commit 0f1dc52e2e3a5cde23ac60ccee39d938e544f26b

vijaysun-omr commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7071,10 +7065,62 @@ void TR_EscapeAnalysis::makeNonContiguousLocalAllocation(Candidate *candidate)
+
+         // generate store to the field
+         const TR::TypeLayoutEntry& fieldEntry = typeLayout->entry(i - 1);
+         TR::SymbolReference* symref = comp()->getSymRefTab()->findOrFabricateShadowSymbol(valueClass,
```
What is the explanation behind creating this fabricated `symref` ?

hzongaro replied:
Are you asking why this `symref` was created but never used?  If so, it looks like that was a mistake on my part in attempting to organize the commits so that the full support for escape analysis of value types accumulated in a reasonably logical fashion.  A later commit in this pull request that adds cold block heapification support for value types actually makes use of the `symref` in a call to [`candidate->findOrSetFieldInfo`](https://github.com/hzongaro/openj9/commit/c2951330f46ea107e6d401e1154d51d3bedfeeef#diff-b5403d98cf7fd820a992fb53dba8c440727fddcd346185442bd1a317c200701cL6781-R6760) and stores it in `field._symRef`.

If you'd like, I can rework these commits so that this `findOrFabricate` call only appears in that later commit.

Or are you looking for additional comments describing the purpose of the call to `findOrFabricateShadowSymbol`?

vijaysun-omr replied:
It was the former I think, i.e. I was wondering why it was introduced in this commit. But you explained that as just an issue with organizing the commits.

hzongaro replied:
Would you like me to rework those two commits so that the call to `findOrFabricateShadowSymbol` only appears in the commit where it's first needed?

vijaysun-omr commented:
What is the status with respect to flattened instances or arrays ? i.e. if I have a flattened field or array element inside an object being considered for stack allocation, will stack allocation be allowed to go ahead ?

vijaysun-omr commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -7371,117 +7365,167 @@ void TR_EscapeAnalysis::heapifyForColdBlocks(Candidate *candidate)
-                        {
-                        translateTT = TR::TreeTop::create(comp(), TR::Node::createCompressedRefsAnchor(heapFieldStore), NULL, NULL);
-                        }
+                     stackFieldLoad = TR::Node::createWithSymRef(newValueHeapificationNode, comp()->il.opCodeForDirectLoad(type), 0, field._symRef);
```
Can this be reached for a flattened field ? If so, will having an auto of that type work ?

hzongaro replied:
All of the symrefs for flattened fields will be the ultimate "leaf" fields, so they will all be of primitive types or reference fields that are not flattened.

hzongaro commented:
> What is the status with respect to flattened instances or arrays ? i.e. if I have a flattened field or array element inside an object being considered for stack allocation, will stack allocation be allowed to go ahead ?

If an object that's created with a `TR::new` instruction has a field of a null-restricted field that is flattened, it is still eligible for stack allocation, as the leaf fields can all be zero initialized.  If it has a null-restricted field that is not flattened, however, as things stand escape analysis is not prepared to create the default instance for the field and will not stack allocate the object.  The check of whether the class is zero initializable happens in [`performAnalysisOnce`](https://github.com/eclipse-openj9/openj9/blob/ecdda12b326693add8c054e99dc82f869a6e22bc/runtime/compiler/optimizer/EscapeAnalysis.cpp#L1087-L1099) and you can find the [implementation in `J9ClassEnv.cpp`](https://github.com/eclipse-openj9/openj9/blob/ecdda12b326693add8c054e99dc82f869a6e22bc/runtime/compiler/env/J9ClassEnv.cpp#L952).

Similarl an array whose component type is a primitive value type will be filtered out from consideration in [`performAnalysisOnce`](https://github.com/eclipse-openj9/openj9/blob/ecdda12b326693add8c054e99dc82f869a6e22bc/runtime/compiler/optimizer/EscapeAnalysis.cpp#L1198-L1209).  That's currently more restrictive than the case of flattenable fields, where the stack allocation can go ahead if the field is flattened.  For flattened arrays, the stack allocation does not currently proceed, despite the fact that the fields of the elements can all be zero initialized.

I can open follow on issues to deal with those cases or, if you'd prefer, I can work on those enhancements under this pull request.

vijaysun-omr commented:
Thanks for the explanation around flattened fields. I am fine with (and probably prefer) working out those other cases as and when the need arises than being proactive at this stage and doing it all especially since we don't have a great sense of relative priority with other things that you could be working on. 

I certainly don't want that extra handling though in this PR, since this PR has more than enough code being committed at once anyway.

vijaysun-omr commented:
I can run tests again once you update the commit with the different diagnostic message you proposed above in response to one of my comments.

hzongaro commented:
> I can run tests again once you update the commit with the different diagnostic message you proposed above in response to one of my comments.

I've updated that diagnostic message.  If you don't mind, I'd like to first squash the more recent commits from review comments before you go ahead with running tests.

vijaysun-omr commented:
Sure, please let me know when it is ready

hzongaro commented:
> Sure, please let me know when it is ready

Thanks, @vijaysun-omr - I've squashed the commits back down to a reasonable set.

vijaysun-omr commented:
jenkins test sanity all jdk17

vijaysun-omr commented:
Checks have passed. Merging.
